### DISCUSSION TITLE
Question about local function declaration and invocation

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#7173] [Q&A]
Logically, a local function should be able to be declared anywhere, but only be invoked after capturing a variable. But the reality is the opposite, it must be declared after capturing a variable, but it can be invoked before the variable declaration. For example, in this code:
```csharp
Hello5();

int i5;

void Hello5()
{
	i5 = 40;
}

Console.WriteLine(i5);
```
The local function `Hello5` captures the variable `i5`, but it is declared after `i5`. However, it can be called before `i5`. How does this work?

**Comment by HaloFour** [comment]
The variable is in scope where the function is lexically declared, but thanks to hoisting the function is in scope through the entire method.  The compiler does normally enforce that the method can't be called until the captures are definitely assigned, but since you are only assigning a value to the variable it's perfectly safe to do that before the variable is definitely assigned.

**Reply by 333fred** [reply]
This particular example seems like a bug. We should error that the local function is called before i5 is declared.

**Reply by HaloFour** [reply]
I didn't know if it was a side effect of hoisting, but I guess that should only apply if `i5` was another local function declaration and not a normal local declaration?
